lib/Rabbit.js

```diff
@@ -0,0 +1,253 @@
+        this._run();
+    }
+
+    async publish(payload) {
+        if (this._connection) {
+            this._config.pubPayload = payload;
+            this._publishedQueueList = [];
+            await this._pubQueue(this._connection);
+            await this._pub(this._connection);
+        } else {
+            await sleep(1000);
+            await this.publish(payload);
+        }
+    }
+
+    async _run() {
+        try {
+            // create connection
+            const connection = await amqp.connect(this._connectionString);
+            this._logger.info("Connected to rmq.");
+            connection.on("error", async (error) => {
+                this._connection = null;
+                this._logger.error("Connection error : ", error);
+                await sleep(1000);
+                await this._run();
+            });
+            connection.on("close", async (error) => {
+                this._connection = null;
+                this._logger.error("Connection close : ", error);
+                await sleep(1000);
+                await this._run();
+            });
+            this._connection = connection;
+            // process pubSub handler
+            await this._subQueue(this._connection);
+            await this._sub(this._connection);
+            await this._pubQueue(connection);
+            await this._pub(connection);
+            // add to channels
+        } catch (error) {
+            this._connection = null;
+            this._logger.warn("Failed to connect to rmq.", error);
+            await sleep(1000);
+            this._logger.info("Attempting to reconnect to rmq.");
+            await this._run();
+        }
+    }
+
+    async _pub(connection) {
+        if (!this._config.pub || !this._connection) {
+            return false;
+        }
+        // extract config
+        const { namespace, pub, pubPayload, serviceName } = this._config;
+        // create channel & publish payload
+        for (const eventName in pub) {
+            const triggerName = pub[eventName];
+            const exchangeName = `${namespace}.${serviceName}.${eventName}`;
+            // if queue is already used for publication, then skip it
+            if (this._publishedQueueList.includes(exchangeName)) {
+                continue;
+            }
+            if ("pubPayload" in this._config) {
+                const content = JSON.stringify(pubPayload);
+                const channel = await connection.createChannel();
+                await channel.assertExchange(exchangeName, "fanout", { durable: true });
+                channel.publish(exchangeName, "", Buffer.from(content), { persistent: true });
+                // note that the queue has already been used for publication
+                this._publishedQueueList.push(exchangeName);
+            }
+        }
+        return true;
+    }
+
+    async _sub(connection) {
+        if (!this._config.sub) {
```

Comment:
I think we should add connection check on this line too

Reply:
correct